ddtrace/internal/datadog/profiling/src/interface.cpp

```diff
@@ -293,7 +293,8 @@ ddup_upload()
         // Rationalize return for interface
         std::cout << "WHOA NOT INITIALIZED" << std::endl;
     }
-    new std::thread(ddup_upload_impl, g_profile); // set it and forget it
+    std::thread(ddup_upload_impl, g_profile).detach(); // set it and forget it
```

Comment:
Is there a way for us to `join` the thread instead? Or would that not be any better than `detach`ing?

Reply:
Ah, that's a good question.  I guess there are scenarios where the average thread lifetime might exceed the upload interval (e.g., if the user specifies a really short upload), but in those scenarios a `join()` could massively slow down the profiler.  At the same time, forceful termination (e.g., `kill`) could work, but I think it bypasses normal thread destruction.  I _think_ that's OK, but I need to see whether it introduces a leak.

Will check it out and try to tighten this up.

Reply:
> I guess there are scenarios where the average thread lifetime might exceed the upload interval (e.g., if the user specifies a really short upload), but in those scenarios a `join()` could massively slow down the profiler.

True that. I think currently we do join the threads on shutdown, but skip it on fork (because the thread would be dead anyway). This of course might be problematic if the child process is short-lived and has produced data that needs to be flushed out. The `join` in this case would add a potentially sizeable delay. However, joining on shutdown in this case would not be any different from what we currently do.

Reply:
One of the interesting things here is that all of the storage is held outside of Python, so if all of the Python threads shut down, then as long as the process group doesn't get forcibly closed (does cpython kill itself after it's joined all of its threads?) these native threads will hang out until they're done uploading.  They won't have to deal with module cleanup.

So, I'd say we need to join on shutdown only if cpython forces us to (because it'll kill our threads otherwise).  I'd need to check that out--do you know?